hibernate-core/src/main/java/org/hibernate/engine/spi/AbstractDelegatingSessionBuilder.java

```diff
@@ -25,19 +25,14 @@
  * @author Gunnar Morling
  * @author Guillaume Smet
  */
-public abstract class AbstractDelegatingSessionBuilder<T extends SessionBuilder> implements SessionBuilder<T> {
+public abstract class AbstractDelegatingSessionBuilder<T extends AbstractDelegatingSessionBuilder<T>> implements SessionBuilder<T> {
 
```

Comment:
The generics type for builder hierarchy serves only one purpose, i.e. a placeholder for its subclass type per se. `SessionBuilder` chosen as the `extends` target is incorrect for it lacks the safety net against the possibility that a generics type not directly under `AbstractDelegatingSessionBuilder` could be used erroneously.
Similar code changes throughout the PR.